File: src/server/common/cursor-handler.ts
import { CustomError } from '../core/errors/base-error';

export interface ListWithCursor<T> {
    list: T[];
    next?: string;
}

export class CursorHandler<T, R extends keyof T> {
    constructor(readonly fields: Array<R>) {}

    parse(cursor: string): Record<R, T[R]> {
        const data = cursor.split(':');
        return data.reduce((acc, item, idx) => {
            const field = this.fields[idx];
            if (!field) throw new CustomError('ER0001', 'Invalid cursor', { cursor, field });
            acc[field] = item as T[R];
            return acc;
        }, {} as Record<R, T[R]>);
    }

    stringify(data: Partial<Record<R, T[R]>>): string {
        return this.fields.map(field => `${data[field]}`).join(':');
    }
}

File: src/swift-server/core/cursor-handler.swift
import Vapor

class CursorHandler<T, R: Hashable> {
	let fields: [R]

	init(_ fields: [R]) {
		self.fields = fields
	}

	func parse(_ cursor: String) throws -> [R: String] {
		let data = cursor.split(separator: ":")
		var result: [R: String] = [:]
		for (index, item) in data.enumerated() {
			guard let field = fields[safe: index] else {
				throw Abort(.badRequest, reason: "Invalid cursor: \(cursor)")
			}
			result[field] = String(item)
		}
		return result
	}

	func stringify(_ data: [R: String]) -> String {
		return fields.compactMap { data[$0] }.joined(separator: ":")
	}
}

struct PageQuery {
    var limit: Int = 100
    var cursor: String? = nil
    
    func getListWithCursor<T>(data: [T], generateCursor: (T)->String) -> ListWithCursor<T> {
        let hasMore = data.count >= limit

        return ListWithCursor(list: data, next: hasMore ? generateCursor(data.last!) : nil)
    }
}
